// react-c3js/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var C3Chart = function (_React$Component) {
  _inherits(C3Chart, _React$Component);

  function C3Chart() {
    _classCallCheck(this, C3Chart);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(C3Chart).apply(this, arguments));
  }

  _createClass(C3Chart, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(newProps) {
      this.updateChart(newProps);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.destroyChart();
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.updateChart(this.props);
    }
  }, {
    key: 'generateChart',
    value: function generateChart(mountNode, config) {
      var c3 = require('c3');
      return c3.generate(_extends({
        bindto: mountNode
      }, config));
    }
  }, {
    key: 'destroyChart',
    value: function destroyChart() {
      try {
        this.chart = this.chart.destroy();
      } catch (err) {
        throw new Error('Internal C3 error', err);
      }
    }
  }, {
    key: 'updateChart',
    value: function updateChart(config) {

      var newConfig = {
        data: config.data,
        size: config.size || {},
        padding: config.padding || {},
        color: config.color || {},
        interaction: config.interaction || {},
        transition: config.transition || {},
        oninit: config.oninit || function () {},
        onrendered: config.onrendered || function () {},
        onmouseover: config.onmouseover || function () {},
        onmouseout: config.onmouseout || function () {},
        onresize: config.onresize || function () {},
        onresized: config.onresized || function () {},
        axis: config.axis || {},
        grid: config.grid || {},
        regions: config.regions || [],
        legend: config.legend || {},
        tooltip: config.tooltip || {},
        subchart: config.subchart || {},
        zoom: config.zoom || {},
        point: config.point || {},
        line: config.line || {},
        area: config.area || {},
        bar: config.bar || {},
        pie: config.pie || {},
        donut: config.donut || {},
        gauge: config.gauge || {}
      };

      if (this.chart) {
        this.destroyChart();
      }

      this.chart = this.generateChart((0, _reactDom.findDOMNode)(this), newConfig);
    }
  }, {
    key: 'render',
    value: function render() {
      var className = this.props.className ? ' ' + this.props.className : '';
      var style = this.props.style ? this.props.style : {};
      return _react2.default.createElement('div', { className: className, style: style });
    }
  }]);

  return C3Chart;
}(_react2.default.Component);

exports.default = C3Chart;
